[Ho Cheok Wai, Sherman] iP

itsnotsherm opened the pull request:
# REX Bot
> "I love it" - Some guy, probably

The REX Bot is a digital rawr-ing companion that helps you remember so you don't have to. It's
- Text-based
- Easy to learn
- ~~FAST~~ SUPER FAST to use
- And it rawrs!

All you need to do is:
1. Download it
2. Double-click it
3. Add your tasks
4. Let Rex manage your tasks for you! 😄 

And it is **FREE**!

Features:
- [x] Managing tasks
- [x] Managing deadlines
- [ ] Reminders (coming soon)

Here's some sample code for how our bot performs the `rawr`:
```java
public class Rawr {
        public static void main(String[] args) {
            System.out.println("rawr");
        }
    }
```
Shoutout to [Rex's favourite website!](https://nus-cs2103-ay2425s1.github.io/website/)

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

itsnotsherm committed:
Rename Duke.java to Rex.java
Rex.java: Add skeletal implementation of Rex chatbot

itsnotsherm committed:
Add implementation of chatbot echo

itsnotsherm committed:
Add Tell Rex custom message during user input

itsnotsherm committed:
Add implementation to store user inputs and display them in a list

itsnotsherm committed:
Add Task class

itsnotsherm committed:
Refactor main method in Rex.java

itsnotsherm committed:
Abstract list command into displayList method

itsnotsherm committed:
Redefine input type to array to extract command from user command

itsnotsherm committed:
Add mark and unmark user command

itsnotsherm committed:
Fix text format for markDone and unmarkDone methods

itsnotsherm committed:
Add ToDo class

itsnotsherm committed:
Add Deadline class

itsnotsherm committed:
Add Event class

itsnotsherm committed:
Add class numberOfTasks variable to track number of tasks created

itsnotsherm committed:
Add skeleton create method for each type of Task

itsnotsherm committed:
Add createTask methods to create respective task types

itsnotsherm committed:
Add createTask method to Rex.java

itsnotsherm committed:
Remove redundant methods

itsnotsherm committed:
Remove redundant factory methods

itsnotsherm committed:
Remove ability to add unknown commands to list, instead echoes unknown commands back

itsnotsherm committed:
Fix text bug and formatting

itsnotsherm committed:
Configure runtest files to compile and run Rex

itsnotsherm committed:
Fix bug where scanner re-initialize every loop, now automated testing can be done

itsnotsherm committed:
Add automated text UI testing using I/O redirection

itsnotsherm committed:
Add help command to list valid commands

itsnotsherm committed:
Add empty list message

itsnotsherm committed:
Add InvalidInputException class

itsnotsherm committed:
Add error handling in switch cases

itsnotsherm committed:
Add InvalidTaskException class

itsnotsherm committed:
Fix split bug in createDeadline and createEvent methods

itsnotsherm committed:
Fix text formatting

itsnotsherm committed:
Add error handling for mark/unmark

itsnotsherm committed:
Add error handling cases

itsnotsherm committed:
Add static modifier to getNumberOfTasks method

itsnotsherm committed:
Add removeTask method

itsnotsherm committed:
Add deleteTask method

itsnotsherm committed:
Add test inputs for delete task feature

itsnotsherm committed:
Add Command enum

itsnotsherm committed:
Integrate Command enum into Rex class

itsnotsherm committed:
Add usage message to command errors

itsnotsherm committed:
Add loadFile method

itsnotsherm committed:
Add updateFile method to update after taskList changes

itsnotsherm committed:
Add formatted method to format text to save file

itsnotsherm committed:
Remove mark/unmark messages

itsnotsherm committed:
Add functionality to load tasks from save file to task list

itsnotsherm committed:
Add data folder to .gitignore

itsnotsherm committed:
Fix text formatting

itsnotsherm committed:
Add text-ui-test data folder to .gitignore

itsnotsherm committed:
Fix rawr bug

itsnotsherm committed:
Change date/time type from String to LocalDateTime

itsnotsherm committed:
Update usage messages to match LocalDateTime input format

itsnotsherm committed:
Add parsers to LocalDateTime to create Deadline and Event tasks

itsnotsherm committed:
Change filepath name to output name of file to rex.txt

itsnotsherm committed:
Update test cases to match new date/time format

itsnotsherm committed:
Add InvalidCommandException

itsnotsherm committed:
Add new classes for OOP

itsnotsherm committed:
Update to throw InvalidCommandException where appropriate

itsnotsherm committed:
Add Parser that processes inputs and files

itsnotsherm committed:
Update task classes to adhere to OOP principles

itsnotsherm committed:
Refactor code to decouple chatbot functionalities in Rex class

itsnotsherm committed:
Fix text format for delete function

itsnotsherm committed:
Refactor classes into packages

itsnotsherm committed:
Add required import statements

itsnotsherm committed:
Add Command import

itsnotsherm committed:
Fix text bug

itsnotsherm committed:
Fix error message bug

itsnotsherm committed:
Merge branch 'add-gradle-support'

itsnotsherm committed:
Refactor switch statements in Command class

itsnotsherm committed:
Add configurations to build

itsnotsherm committed:
Enable assertions

itsnotsherm committed:
Add JUnit tests for task-related classes

itsnotsherm committed:
Add META-INF to .gitignore

itsnotsherm committed:
Add find functionality into Rex

itsnotsherm committed:
Add JavaDoc header comments for classes and its methods

itsnotsherm committed:
Refactor code to match Java Coding Standard

itsnotsherm committed:
Resolve merge conflicts involving JavaDocs and refactored code

itsnotsherm committed:
Fix empty list bug

itsnotsherm committed:
Resolve merge conflicts and add find task functionality

LiuYJ2002 commented on src/main/java/rex/Rex.java:
```diff
@@ -0,0 +1,151 @@
+ * It integrates storage, task management, and user interaction functionalities.
+ */
+public class Rex {
+    private Storage storage;
+    private TaskList taskList;
+    private Ui ui;
```
I like your naming. Should taskList be in plural?

itsnotsherm replied:
Thank you for the suggestion, will keep that in mind when making changes to my code :)

LiuYJ2002 commented on src/main/java/rex/task/Deadline.java:
```diff
@@ -0,0 +1,48 @@
+     * @return A formatted string representing the Deadline task.
+     */
+    @Override
+    public String formatted() {
```
Would it be better if its named as formatter?

itsnotsherm replied:
It would be better, thank you for pointing that out!

LiuYJ2002 commented on src/main/java/rex/task/Task.java:
```diff
@@ -0,0 +1,87 @@
+    /**
+     * Marks the task as done by setting its completion status to true.
+     */
+    public void markDone() {
```
Would it be better if its markAsDone?

itsnotsherm replied:
I believe so, will make the necessary changes!

LiuYJ2002 commented on src/main/java/rex/task/TaskList.java:
```diff
@@ -0,0 +1,250 @@
+ */
+public class TaskList {
+    private Storage storage;
+    private ArrayList<Task> list;
```
Would it be better if list is in plural?

itsnotsherm replied:
Nice spot, will rename it to tasks instead for clarity!

LiuYJ2002 reviewed (commented):
LGTM just a few nits to be fixed

nuagetsu commented on src/main/java/rex/task/Task.java:
```diff
@@ -2,7 +2,7 @@
 public class Task {
     private static int numberOfTasks = 0;
-    private String description;
+    private final String description;
```
Should the constant be in screaming snake case instead?

itsnotsherm replied:
Thank you for the suggestion, I believe the description is actually not a constant, as it is first assigned a String when a Task constructor is called! Constants would be in the form static final instead :)

nuagetsu commented on src/main/java/rex/util/Parser.java:
```diff
@@ -16,6 +16,8 @@
  * It includes methods to parse commands, tasks, and dates, as well as handle input validation.
```
Should the javadocs comment start with indicative mood ie "Includes" rather than "It", as the it is already inferred to be the class?

itsnotsherm replied:
Thanks for this suggestion, I believe the current comment should be okay but I will take it into consideration :)

nuagetsu commented on src/main/java/rex/util/Ui.java:
```diff
@@ -13,14 +13,20 @@
  * displaying messages, prompts, and error messages.
```
Perhaps could start with a capital letter for consistency, as some comments start with capital letters while others start with lower case

itsnotsherm replied:
Nice spot, thank you!

nuagetsu commented on src/main/java/rex/util/Storage.java:
```diff
@@ -45,11 +46,12 @@ public void loadFile(TaskList list) throws IOException {
-        // Delete the temp file if it exists and create a new temp file
-        Files.deleteIfExists(Paths.get(temppath));
+
+        // Create temp file to copy taskList from
```
Would it be better to start with the indicative mood, ie "Creates" rather than "Create" for standardization? I noticed that in several comments it is started with "Create" while others use the indicative mood.

itsnotsherm replied:
Thank you for pointing this out, will make the necessary changes!

nuagetsu reviewed (commented):
Overall LGTM, just a few nits for coding standard and javadocs conventions.

itsnotsherm committed:
Add checkstyle plugin

itsnotsherm committed:
Add config folder to .gitignore

itsnotsherm committed:
Refactor code to adhere to checkstyle and PR review comments

itsnotsherm committed:
Add dependencies for JavaFX

itsnotsherm committed:
Add dependencies for JavaFX

itsnotsherm committed:
Add GUI prototype using JavaFX

itsnotsherm committed:
Integrate chatbot logic into newly implemented GUI

itsnotsherm committed:
Refactor code to pass checkstyle

itsnotsherm committed:
Add error message for missing keyword in find command

itsnotsherm committed:
Add assert statements at various points in code

Code should include some assertions to help us indicate any possible
bug in the code caused by mistakes by the developer.

Let's use assertions in our code to check for the following
(but not limited to):
* Null checks
* Input validation
* Existence of file for parsing

itsnotsherm committed:
Rewrite TaskList methods to use Java Streams

itsnotsherm committed:
Add commandToOutput method in Rex that uses varargs

itsnotsherm committed:
Refactor code to improve code quality

The way some of the code is written diminishes its readability, may be
prone to errors and contain unnecessary and heavy commenting.

Let's,
* Avoid complicated expressions in code, such as nested parentheses
* Avoid long methods by extracting into separate methods
* Lay out code such that it adheres to a logical flow and structure
* Use default branches in case statements to handle all cases
* Delete dead code the moment it becomes redundant

itsnotsherm committed:
Merge branch 'branch-A-Varargs'

itsnotsherm committed:
Merge pull request #1 from itsnotsherm/branch-A-CodeQuality

Refactor code to improve code quality

itsnotsherm committed:
Add assert statements at various points in code

Code should include some assertions to help us indicate any possible bug
in the code caused by mistakes by the developer.

Let's use assertions in our code to check for the following (but not
limited to):
* Null checks
* Input validation
* Existence of file for parsing

itsnotsherm committed:
Rewrite TaskList methods to use Java Streams

itsnotsherm committed:
Create gradle.yml file to set up CI

itsnotsherm committed:
Rewrite code to pass checkstyle

itsnotsherm committed:
Add ability to view schedule on specific date

itsnotsherm committed:
Add config file for checkstyle

itsnotsherm committed:
Add error handling where task description is whitespaces only

itsnotsherm committed:
Add error handling where event from date/time is after to

itsnotsherm committed:
Add class ParserTest

itsnotsherm committed:
Decouple filepath in storage class

itsnotsherm committed:
Merge branch 'branch-A-MoreTesting'

itsnotsherm committed:
Improve GUI using JavaFX and CSS

itsnotsherm committed:
Change background colour

itsnotsherm committed:
Add product screenshot

itsnotsherm committed:
Change format of date and time for clarity

itsnotsherm committed:
Update README to provide user guide

itsnotsherm committed:
Add and edit header comments to adhere to coding standard

itsnotsherm committed:
Edit code to adhere to checkstyle

itsnotsherm committed:
Add command table summary to user guide

itsnotsherm committed:
Update file path of test files in TaskListTest
